### ISSUE TITLE
[@nestjs/microservice] EventPattern could be limited to specific queue

### ISSUE THREAD

**Issue by Dzixxx** [nestjs/nest#10216]
### Is there an existing issue that is already proposing this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe it

We have few applications with two Nats setups:
```
const opts = {
  servers: [ 'a' ],
  serializer: new MyNatsSerializer()
  deserializer: new MyNatsDeserializer()
}

// this will behave like request/reply (queue)
app.connectMicroservice({
  transport: Transport.NATS,
  options: {
    ...opts,
    queue: 'my-lovely-queue'
  }
})

// this will behave like pubsub
app.connectMicroservice({
  transport: Transport.NATS,
  options: {
    ...opts
  }
})
```

and we have an EventPattern('hello') which should be triggered only once (pubsub) but because of two subscriptions (two setups above) we're receiving that message twice (sometimes once) because one is from pubsub and second from queue (depends which subscription will be picked up by queue and to which app it will go)
 

### Describe the solution you'd like

Would be perfect if we could somehow make event pattern parametrized to execute only for specific queue or to be handled by generic client without queue provided in setup:
```
@EventPattern('hello')
sayHi(data) {
  // I should receive message handled by 1st and 2nd connectMicroservice server
}

@EventPattern('hello1', { queue: false })
sayHi1(data) {
  // I should receive message handled by 2nd connectMicroservice server
}

@EventPattern('hello2', { queue:'my-lovely-queue' })
sayHi2(data) {
  // I should receive message handled by 1st connectMicroservice server
}
```

inside server implementation I think there should be possibility to make this kind of if statement to check if those extras from reflect are matching server options[queue] ;> 

### Teachability, documentation, adoption, migration strategy

Extending NATS section docs that event pattern can be limited to specific queue, otherwise it will be triggered as many times as in default mode (default NATS behavior)

### What is the motivation / use case for changing the behavior?

To use NATS in pubsub and request/reply way at the same time with only one method execution (triggered by one server instance) in single app

**Comment by Dzixxx** [comment]
In NATS documentation there is a clear statement that subscriptions for queue are separate subscriptions so that's why I can have 2 method execution on single application (https://github.com/nats-io/nats.js#queue-groups)

**Comment by Dzixxx** [comment]
I overcome problem with two custom setups 
```
const PUB_SUB_NATS = Symbol('nats-pub-sub');
const QUEUE_NATS = Symbol('nats-queue');

// ctrl
@MessagePattern('hi', QUEUE_NATS)
sayHi() {
 // nevermind
}

@EventPatter('hello, PUB_SUB_NATS)
sayHi() {
 // nevermind
}

// setup

class NatsPubSubServer extends ServerNats {
  constructor(opts) {
    super(opts);
    this.transportId = PUB_SUB_NATS;
  }
}

class NatsQueueServer extends ServerNats {
  constructor(opts) {
    super(opts);
    this.transportId = QUEUE_NATS;
  }
}

// ...
app.connectMicroservice({
  transport: PUB_SUB_NATS,
  strategy: new NatsPubSubServer({
    // same options as before
  })
}, {
  inheritAppConfig: true
});

app.connectMicroservice({
  transport: QUEUE_NATS,
  strategy: new NatsQueueServer({
    // same options as before
  })
}, {
  inheritAppConfig: true
});

```

In this way I'm able to create as many NatsSevers (or clients in fact) as I want and limit EventPatterns/MessagePatterns to specific transport type (Symbol) only. Hope it will help others ;)

**Comment by Dzixxx** [comment]
If this approach is ok, please close the ticket (otherwise -> are U able to show direction which I should go to implement it in better way) ;)

**Comment by kamilmysliwiec** [comment]
Perfect usage of the custom strategy feature, @Dzixxx! If you want & have time, I'd suggest posting an article on this!

**Comment by caiuscitiriga-zupit** [comment]
Hi! @Dzixxx 
I've tried to implement your setup, but without success. For example typescript complains about the fact that `transportId` is readonly internally, and even by bypassing that with an `any`, once I run the apps the queues don't work at all, and I receive double messages/events on all the clients 😭 

 Could you share some actual code please 🙏 ?

**Comment by caiusCitiriga** [comment]
Hi! @Dzixxx 
I've tried to implement your setup, but without success. For example typescript complains about the fact that transportId is readonly internally, and even by bypassing that with an any, once I run the apps the queues don't work at all, and I receive double messages/events on all the clients 😭

Could you share some actual code please 🙏 ?

**Comment by caiusCitiriga** [comment]
@kamilmysliwiec 

I've came up with this extension of the `ServerNats` class provided by Nest. 

TLDR; bypasses the `queue` property in the `NatsOptions`, and takes a list of patterns that SHOULD be queued under the given `queue` param.

## main.ts

```ts
import { Logger } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { Transport, ServerNats, NatsOptions } from '@nestjs/microservices';
import { Client } from '@nestjs/microservices/external/nats-client.interface';

import { AppModule } from './app.module';
import { MessagingTransportBuilder } from './core/messaging-transport/messaging-transport-builder';

export const PATTERNS = {
  events: {
    helloSimple: 'event.get-hello',
    helloComplex: { event: 'get-hello', foo: 'bar' },
  },
  messages: {
    helloSimple: 'message.get-hello',
    helloComplex: { message: 'get-hello', foo: 'bar' },
  },
};

class NatsServerWithExclusiveQueues extends ServerNats {
  private readonly stringPatterns: string[] = [];

  constructor(
    private readonly patterns: unknown[],
    private readonly queue: string,
    private readonly opts: NatsOptions,
  ) {
    super(opts);
    this.patterns = this.patterns.map((p) => {
      if (typeof p !== 'object') return p;

      const orderedPatternKeys = Object.keys(p)
        .sort()
        .reduce((res, currKey) => {
          res[currKey] = p[currKey];
          return res;
        }, {});
      return orderedPatternKeys;
    });

    this.stringPatterns = this.patterns.map((p) =>
      typeof p === 'string' ? p : JSON.stringify(p),
    );
  }

  public override bindEvents(client: Client) {
    const subscribe = (channel: string) => {
      Logger.log(
        `[${channel}]: ${
          this.stringPatterns.includes(channel) ? 'QUEUED' : 'NOT QUEUED'
        }`,
      );
      client.subscribe(channel, {
        callback: this.getMessageHandler(channel).bind(this),
        queue: this.stringPatterns.includes(channel) ? this.queue : undefined,
      });
    };

    const registeredPatterns = [...this.messageHandlers.keys()];
    registeredPatterns.forEach((channel) => subscribe(channel));
  }
}

async function bootstrap() {
  const app = await NestFactory.createMicroservice(AppModule, {
    transport: Transport.NATS,
    strategy: new NatsServerWithExclusiveQueues(
      [PATTERNS.messages.helloSimple, PATTERNS.messages.helloComplex],
      'queue',
      MessagingTransportBuilder.build() as NatsOptions,
    ),
  });

  await app.listen();
}
bootstrap();
```

## Controller

```ts
```

**Comment by Dzixxx** [comment]
> I've tried to implement your setup, but without success. For example typescript complains about the fact that transportId is readonly internally, 

@caiusCitiriga try in this way (I'm not saying that it's correct with TS but works for me ;>) 
```
const EXCLUSIVE = Symbol('VALUE_FOR_EXCLUSIVE_QUEUE')
```
use it when extending ServerNats class
```
// ... constructor
(this['transportId'] as unknown) = EXCLUSIVE
```
when connecting microservice use that value as transport
```
app.connectMicroservice({
  transport: EXCLUSIVE,
  strategy: new YourExtendedServerNats({
    // ... configuration, serializer, deserializer etc
    queue: NameOfYourQueue
  })
})
```
then 
```
@EvenPattern(PATTERNS.events.helloSimple, EXCLUSIVE)
```
